lib/cea/mp4_cea_parser.js

```diff
@@ -0,0 +1,229 @@
+
+    /**
+     * Default sample duration, as specified by the TREX box.
+     * @private {!number}
+     */
+    this.defaultSampleDuration_ = 0;
+
+    /**
+     * Default sample size, as specified by the TREX box.
+     * @private {!number}
+     */
+    this.defaultSampleSize_ = 0;
+  }
+
+  /**
+   * Parses the init segment. Gets Default Sample Duration and Size from the
+   * TREX box, and constructs a map of Track IDs to timescales. Each TRAK box
+   * contains a track header (TKHD) containing track ID, and a media header box
+   * (MDHD) containing the timescale for the track
+   * @override
+   */
+  init(data) {
+    const Mp4Parser = shaka.util.Mp4Parser;
+    const trackIds = [];
+    const timescales = [];
+
+    new Mp4Parser()
+        .box('moov', Mp4Parser.children)
+        .box('mvex', Mp4Parser.children)
+        .fullBox('trex', (box) => {
+          const parsedTREXBox = shaka.util.Mp4BoxParsers.parseTREX(
+              box.reader);
+
+          this.defaultSampleDuration_ = parsedTREXBox.defaultSampleDuration;
+          this.defaultSampleSize_ = parsedTREXBox.defaultSampleSize;
+        })
+        .box('trak', Mp4Parser.children)
+        .fullBox('tkhd', (box) => {
+          goog.asserts.assert(
+              box.version != null,
+              'TKHD is a full box and should have a valid version.');
+          const parsedTKHDBox = shaka.util.Mp4BoxParsers.parseTKHD(
+              box.reader, box.version);
+          trackIds.push(parsedTKHDBox.trackId);
+        })
+        .box('mdia', Mp4Parser.children)
+        .fullBox('mdhd', (box) => {
+          goog.asserts.assert(
+              box.version != null,
+              'MDHD is a full box and should have a valid version.');
+          const parsedMDHDBox = shaka.util.Mp4BoxParsers.parseMDHD(
+              box.reader, box.version);
+          timescales.push(parsedMDHDBox.timescale);
+        })
+        .parse(data, /* partialOkay= */ true);
+
+    // At least one track should exist, and each track should have a
+    // corresponding Id in TKHD box, and timescale in its MDHD box
+    if (trackIds.length <= 0 || timescales.length <= 0 ||
+      trackIds.length != timescales.length) {
+      throw new shaka.util.Error(
+          shaka.util.Error.Severity.CRITICAL,
+          shaka.util.Error.Category.TEXT,
+          shaka.util.Error.Code.INVALID_MP4_CEA);
+    }
+
+    // Populate the map from track Id to timescale
+    trackIds.forEach((trackId, idx) => {
+      this.trackIdToTimescale_.set(trackId, timescales[idx]);
+    });
+  }
+
+  /**
+   * Parses each video segment. In fragmented MP4s, MOOF and MDAT come in
+   * pairs. The following logic gets the necessary info from MOOFs to parse
+   * MDATs (base media decode time, sample sizes/offsets/durations, etc),
+   * and then parses the MDAT boxes for CEA-708 packets using this information.
+   * CEA-708 packets are returned in the callback.
+   * @override
+   */
+  parse(data, on708Data) {
```

Comment:
If this happens synchronously, why is on708Data a callback?  Why wouldn't we just return the results instead?

Reply:
In each call to parse(data, on708Data), we can have multiple SEI messages that contain caption packets in the "data" parameter. If we were to return these results as you said, we would have to buffer these results in an array, and then return them to the caller (unless we used a generator or something of that sort). However, this "buffering" of results can be avoided if we use this callback. Whenever we see the result, just pass it in the callback. Although the implementation of the callback on the caller side isn't yet fully implemented (You will see the Todo comment that I left), the caller will actually parse this data immediately in a decoder, and no data will be buffered.

This will in turn save us memory and unnecessary buffering on the client side. Let me know what you think of this, and if you have any suggestions! :)

Reply:
update: changed this to return results as an array.